central/declarativeconfig/updater/auth_provider_updater.go

```diff
@@ -75,18 +83,37 @@ func (u *authProviderUpdater) DeleteResources(ctx context.Context, resourceIDsTo
 	var authProviderDeletionErr *multierror.Error
 	var authProviderIDs []string
 	for _, authProvider := range authProviders {
-		if err := u.authProviderRegistry.DeleteProvider(ctx, authProvider.GetId(), true, true); err != nil {
-			authProviderDeletionErr = multierror.Append(authProviderDeletionErr, err)
-			authProviderIDs = append(authProviderIDs, authProvider.GetId())
-
-			u.reporter.UpdateIntegrationHealthAsync(utils.IntegrationHealthForProtoMessage(authProvider, "", err,
-				u.idExtractor, u.nameExtractor))
+		referencingGroups, err := u.groupDS.GetFiltered(ctx, func(group *storage.Group) bool {
+			return group.GetProps().GetAuthProviderId() == authProvider.GetId()
+		})
+		if err != nil {
+			authProviderDeletionErr, authProviderIDs = u.processDeletionError(authProviderDeletionErr, err, authProviderIDs, authProvider)
 			continue
 		}
-		// TODO(ROX-14700): This currently also deletes imperative groups and should resolve these references instead.
-		if err := u.groupDS.RemoveAllWithAuthProviderID(ctx, authProvider.GetId(), true); err != nil {
-			log.Errorf("Error deleting groups for auth provider id %s: %v", authProvider.GetId(), err)
+		for _, group := range referencingGroups {
+			ctxToUse := utils2.IfThenElse(declarativeconfig.IsDeclarativeOrigin(group.GetProps()), ctx, deleteImperativeGroupsCtx)
+			if err = u.groupDS.Remove(ctxToUse, group.GetProps(), true); err != nil {
+				authProviderDeletionErr, authProviderIDs = u.processDeletionError(authProviderDeletionErr, err, authProviderIDs, authProvider)
+			}
+		}
+		// If err != nil, that means that one of the groups was not deleted.
+		// Since group is valid only if it references existing auth provider, we can't delete auth provider
+		// until all referencing groups are successfully deleted.
+		if err != nil {
+			continue
+		}
+		if err := u.authProviderRegistry.DeleteProvider(ctx, authProvider.GetId(), true, true); err != nil {
+			authProviderDeletionErr, authProviderIDs = u.processDeletionError(authProviderDeletionErr, err, authProviderIDs, authProvider)
 		}
 	}
 	return authProviderIDs, authProviderDeletionErr.ErrorOrNil()
 }
+
+func (u *authProviderUpdater) processDeletionError(authProviderDeletionErr *multierror.Error, err error, authProviderIDs []string, authProvider *storage.AuthProvider) (*multierror.Error, []string) {
+	authProviderDeletionErr = multierror.Append(authProviderDeletionErr, err)
```

Comment:
Looking at this, should we create errors _specifically_ for auth providers?

Currently, in case you had two auth providers created declaratively and both would be deleted, this would lead to a single error being reuse for both integration healths, maybe we should ensure that the error will only include actual errors for the specific auth provider?